include/sparrow/buffer_adaptor.hpp

```diff
@@ -0,0 +1,499 @@
+
+        buffer_adaptor(buffer_reference& buf)
+            : _buffer(buf)
+        {
+        }
+
+        constexpr pointer data() noexcept;
+        constexpr const_pointer data() const noexcept;
+
+
+        // Element access()
+
+        constexpr reference operator[](size_type idx);
+        constexpr const_reference operator[](size_type idx) const;
+
+        constexpr reference front();
+        constexpr const_reference front() const;
+
+        constexpr reference back();
+        constexpr const_reference back() const;
+
+        // Iterators
+
+        constexpr iterator begin() noexcept;
+        constexpr iterator end() noexcept;
+
+        constexpr const_iterator begin() const noexcept;
+        constexpr const_iterator end() const noexcept;
+
+        constexpr const_iterator cbegin() const noexcept;
+        constexpr const_iterator cend() const noexcept;
+
+        constexpr reverse_iterator rbegin() noexcept;
+        constexpr reverse_iterator rend() noexcept;
+
+        constexpr const_reverse_iterator rbegin() const noexcept;
+        constexpr const_reverse_iterator rend() const noexcept;
+
+        constexpr const_reverse_iterator crbegin() const noexcept;
+        constexpr const_reverse_iterator crend() const noexcept;
+
+        // Capacity
+
+        [[nodiscard]] constexpr size_type size() const noexcept;
+        [[nodiscard]] constexpr size_type max_size() const noexcept;
+        [[nodiscard]] constexpr size_type capacity() const noexcept;
+        [[nodiscard]] constexpr bool empty() const noexcept;
+        constexpr void reserve(size_type new_cap);
+        constexpr void shrink_to_fit();
+
+        // Modifiers
+
+        constexpr void clear() noexcept;
+
+        constexpr iterator insert(const_iterator pos, const value_type& value);
+        constexpr iterator insert(const_iterator pos, size_type count, const value_type& value);
+        template <class InputIt>
+            requires std::input_iterator<InputIt>
+        constexpr iterator insert(const_iterator pos, InputIt first, InputIt last);
+        constexpr iterator insert(const_iterator pos, std::initializer_list<value_type> ilist);
+
+        template <class... Args>
+        constexpr iterator emplace(const_iterator pos, Args&&... args);
+
+        constexpr iterator erase(const_iterator pos);
+        constexpr iterator erase(const_iterator first, const_iterator last);
+
+        constexpr void push_back(const value_type& value);
+
+        constexpr void pop_back();
+
+        constexpr void resize(size_type new_size);
+        constexpr void resize(size_type new_size, const value_type& value);
+
+    private:
+
+        static constexpr double _sizeof_ratio = static_cast<double>(sizeof(buffer_reference_value_type))
+                                                / static_cast<double>(sizeof(value_type));
+
+        static constexpr auto _type_size = static_cast<size_type>(1. / _sizeof_ratio);
```

Comment:
What do you think of the names `from_to_size_ratio` and `to_from_size_ratio`?
Also why having a `size_type` and not a double here?

Reply:
Good idea